src/functions_math.jl

```diff
@@ -0,0 +1,57 @@
+# This supports nonbroadcasting math on NamedDimArrays
+
+for op in (:+, :-)
+    @eval function Base.$op(a::NamedDimsArray{A}, b::NamedDimsArray{B}) where {A,B}
+        L = combine_names(A, B)
+        data = $op(parent(a), parent(b))
+        return NamedDimsArray{L}(data)
+    end
+    @eval function Base.$op(a::NamedDimsArray{L}, b::AbstractArray) where {L}
+        data = $op(parent(a), b)
+        return NamedDimsArray{L}(data)
+    end
+    @eval function Base.$op(a::AbstractArray, b::NamedDimsArray{L}) where {L}
+        data = $op(a, parent(b))
+        return NamedDimsArray{L}(data)
+    end
+end
+
+
+# Matrix product
+matrix_prod_error(A, B) = throw(DimensionMismatch(
```

Comment:
Good, yes, better.

I also did not like the thing I had.
It is an utter abuse of the type system.
Creating so called Tuple types, that can't ever be instantiated.

We can get rid of it now,
since we are handling the error case seperately.

```
Base.@pure function check_matmul_dims(a::Tuple{Vararg{Symbol}}, b::Tuple{Vararg{Symbol}})
    a_dim = a[end]
    b_dim = b[1]

    if a_dim !== b_dim && a_dim !== :_ && b_dim !== :_
        throw(DimensionMismatch(
            "Cannot take matrix product of arrays with different inner dimension names. $a vs $b"
        ))
    end

    return nothing
end


matmul_names((a1,a2),(b,))=(a1,b)
matmul_names((a1,a2),(b1,b2))=(a1,b2)
matmul_names((a1,),(b1,b2))=(a1,b2)

function determine_matmul_names(names_a, names_b)
    check_matmul_dims(names_a, names_b)
    return matmul_names(names_a, names_b)
end
```

And that is fully performant. Onlty allocating the returned tuple
It generate the code that I say should constant propegate,

```

julia> @btime (()->determine_matmul_names((:foo,), (:foo,:bar)))()
  7.217 ns (1 allocation: 32 bytes)
(:foo, :bar)

julia> @code_typed (()->determine_matmul_names((:baz,:foo), (:foo,:bar)))()
CodeInfo(
1 ─ %1 = (Base.getfield)((:baz, :foo), 1)::Symbol
│   %2 = (Base.getfield)((:foo, :bar), 2)::Symbol
│   %3 = (Core.tuple)(%1, %2)::Tuple{Symbol,Symbol}
└──      return %3
) => Tuple{Symbol,Symbol}
```

One problem.
Your so called `@pure` function is not pure.
String interpolation, calls `print` under the hood.
Print is a _generic function_.
`@pure` functions are not allowed to call _generic_functions_

Reply:
The following checks at compile time,
and generates perfect code_typed

```
function valid_matmul_dims(a::Tuple{Vararg{Symbol}}, b::Tuple{Vararg{Symbol}})
    a_dim = a[end]
    b_dim = b[1]

    return a_dim === b_dim || a_dim === :_ || b_dim === :_
end


matmul_names((a1,a2),(b,))=(a1,b)
matmul_names((a1,a2),(b1,b2))=(a1,b2)
matmul_names((a1,),(b1,b2))=(a1,b2)

function determine_matmul_names(names_a, names_b)
    valid_matmul_dims(names_a, names_b) || throw(DimensionMismatch(
            "Cannot take matrix product of arrays with different inner dimension names. $a vs $b"
        ))
    return matmul_names(names_a, names_b)
end
```

see the lack of exception handling:
```
julia> @btime (()->determine_matmul_names((:foo,), (:foo,:bar)))()
  7.340 ns (1 allocation: 32 bytes)
(:foo, :bar)

julia> @btime (()->determine_matmul_names((:baz,:foo), (:foo,:bar)))()
  7.322 ns (1 allocation: 32 bytes)
(:baz, :bar)

julia> @code_typed (()->determine_matmul_names((:baz,:foo), (:foo,:bar)))()
CodeInfo(
1 ─ %1 = (Base.getfield)((:baz, :foo), 1)::Symbol
│   %2 = (Base.getfield)((:foo, :bar), 2)::Symbol
│   %3 = (Core.tuple)(%1, %2)::Tuple{Symbol,Symbol}
└──      return %3
) => Tuple{Symbol,Symbol}
```

Reply:
The old version could identify the new names at compile time, but the new version can't. I think we still need the tuple type hack (at least my reduced version).

Reply:
Then we do as follows: and put this everywhere we have marked as allocating once.

```
function determine_matmul_names(a, b)
    valid_matmul_dims(a, b) || throw(DimensionMismatch(
            "Cannot take matrix product of arrays with different inner dimension names. $a vs $b"
        ))
    res = matmul_names(a, b)
    compile_time_return_hack(res)
end 

_compile_time_return(::Val{X}) where X = X
compile_time_return_hack(x) = _compile_time_return(Val{x}())
```

Which works,
it is just stupid that it works:

```

julia> @btime (()->determine_matmul_names((:foo,), (:foo,:bar)))()
  1.488 ns (0 allocations: 0 bytes)
(:foo, :bar)

julia> @btime (()->determine_matmul_names((:baz,:foo), (:foo,:bar)))()
  1.246 ns (0 allocations: 0 bytes)
(:baz, :bar)

julia> @code_typed (()->determine_matmul_names((:baz,:foo), (:foo,:bar)))()
CodeInfo(
1 ─     return (:baz, :bar)
) => Tuple{Symbol,Symbol}
```

Reply:
Thanks I hate it 👍